src/libasr/modfile.cpp

```diff
@@ -57,6 +57,34 @@ std::string save_modfile(const ASR::TranslationUnit_t &m) {
     return b.get_str();
 }
 
+std::string save_pycfile(const ASR::TranslationUnit_t &m) {
+#ifdef WITH_LFORTRAN_BINARY_MODFILES
+    BinaryWriter b;
+#else
+    TextWriter b;
+#endif
+    // Header
+    b.write_string(lfortran_modfile_type_string);
+    b.write_string(LFORTRAN_VERSION);
+
+    // AST section: Original module source code:
+    // Currently empty.
+    // Note: in the future we can save here:
+    // * A path to the original source code
+    // * Hash of the orig source code
+    // * AST binary export of it (this AST only changes if the hash changes)
+
+    // ASR section:
+
+    // Export ASR:
+    // Currently empty.
+
+    // Full ASR:
+    b.write_string(serialize(m));
+
+    return b.get_str();
+}
```

Comment:
I would merge this with LFortran --- I think all LCompilers can then reuse it, just change the extension. I think there is nothing special about LPython's `.pyc` compared to LFortran's `.mod`, I think they can be exactly identical. It's just ASR that is saved, it's independent of the frontend.

Reply:
There is one thing different though. Fortran files have specific module symbols. Only those are saved in the `.mod` files. However in the case Python, whole file can be a module. We just have to import something from it and it becomes a module. But if you call it using `python` command then it acts as an "executable". So what I have done is if any file is compiled with `-c` option of LPython then its ASR (i.e., the full `ASR::TranslationUnit_t`) gets saved in a `.pyc` file. Does that make sense?

Reply:
We can discuss it more. What I had in mind is that a single Python file is exactly 100% equivalent to a single Fortran module file, at the ASR level.

Reply:
(Well, there is a difference that Python modules can be nested, but this is something that people have requested for Fortran also to do, so we should allow that at the ASR level in some clean way.)

Reply:
Cool. Let me try out implementing the rest of the steps I have written in https://github.com/lcompilers/lpython/pull/999#issue-1345166526. Let's see if it works till the end.